File: MATLAB/src/gPC Expansion/gpc_covariance.m
function cov = gpc_covariance(Norm_psi,x_c, y_c)
% gpc_covariance: This function computes covariance between two random
%                 processes (x and y), given the coefficients of gPC 
%                 expansion (x_c and y_c) of both x and y.
%                 Instead, if y_c is not provided or empty, the function 
%                 computes the auto-covariance of random process x.
%
% Inputs:
%       (i) Norm_psi = gPC Normalization coefficients (1 x N array)
%       (ii) x_c = gPC coefficients of random process x (1 x N array)
%       (iii) y_c = gPC coefficients of random process y (1 x N array)
%
% Output:
%        cov = covariance between x and y if nargin=3 (2 x 2 array), autocovariance of x
%              if nargin = 2 (scalar)

%% Covariance calculation

% Covariance is computed as a normalization weighted inner product between
% 2nd to Nth order gPC coefficients (excluding the zeroth order)

% if y_c is not specified, equate y_c to x_c to compute autocovariance of x
if nargin<3 || isempty(y_c)
    y_c = x_c;
end

% If x_c or y_c is row vector, convert it into column vector;
if iscolumn(x_c)
    x_c = x_c';
end

if iscolumn(y_c)
    y_c = y_c';
end

% Extract 2nd to Nth order gPC coefficients (excluding the zeroth order)
x_c_req = x_c(:,2:end);
y_c_req = y_c(:,2:end);

% Diagonal Normalization matrix (N-1 x N-1 array)
Gamma_req = diag(Norm_psi(2:end));

% Weighted inner product 
cov = x_c_req*Gamma_req*y_c_req';

File: gPC-Python-Toolbox/gPC Expansion/gpc_covariance.py
# gpc_covariance: This function computes covariance between two random
#                 processes (a and b), given the coefficients of gPC 
#                 expansion (a_c and b_c ) of both a and b.
#                 Instead, if b_c is not provided or empty, the function 
#                 computes the auto-covariance of random process a.

# Inputs: (variable number of arguments can be passed from calling environment in arg)
#       (i) nrm = gPC Normalization coefficients (1 x N array)
#       (ii) a_c = gPC coefficients of random process a (1 x N array)
#       (iii) b_c = gPC coefficients of random process b (1 x N array)

# Output:
#        covv = covariance between a and b if coefficient of gPC expansion for both
#               a and b is provided (2 x 2 array), autocovariance of a if 
#               b_c is not provided (scalar)


import numpy as np
def gpc_covariance(*arg):
    nrm = arg[0]
    
    a_c = arg[1]
    
    # If b_c is not specified, we equate b_c = a_c and compute the (auto-)
    # covariance of a

    if (len(arg)<3):      # b_c and a_c must be numpy array.
        b_c = a_c
    else:
        b_c = arg[2]
    
    # Weighted inner product
            
    #        i) a_c[0:,1:] and b_c[0:,1:] are extraction of 2nd to Nth order
    #           gPC coefficients
    #        ii) np.diag(nrm[0,1:]) gives diagonal Normalization matrix (N-1 x N-1 array)   
   
    if a_c.shape[0]==2:
     
        covv = np.dot(np.dot(a_c[0:,1:],np.diag(nrm[0,1:])),
                  np.transpose(b_c[0:,1:]))
    else:
        covv = np.dot(np.dot(a_c[1:],np.diag(nrm[0,1:])),
                  np.transpose(b_c[1:]))
    return(covv)



